Sources/MongoSwift/Credential.swift

```diff
@@ -19,7 +19,7 @@ internal struct Credential: Decodable, Equatable {
 }
 
 /// Possible authentication mechanisms.
-internal enum AuthMechanism: String, Decodable {
+public enum AuthMechanism: String, Decodable {
```

Comment:
for forward compatibility we might want to add an `.other(String)` case. this would mean we don't have to make a breaking change if a new mechanism is added.
I don't have a good sense of how often we add auth mechanisms though - thoughts @patrickfreed @mbroadst ?

Reply:
maybe we should give this the full name now, i.e. `AuthenticationMechanism`

Reply:
Is it breaking to add a field to an enum? Would it make sense to leave out the flexibility of `.other(String)` to force an update for something as major as a new mechanism, basically making sure the driver supports it correctly?
Perhaps we could move the enum inside the Credential struct giving us  the `Credential.Mechanism` type?

Reply:
> Is it breaking to add a field to an enum?

adding a new case to an enum is technically breaking, yes, as any `switch` statements that were previously exhaustive are no longer exhaustive.
admittedly it's unlikely that someone would be switching on auth mechanism in their own code but it's possible.

> Would it make sense to leave out the flexibility of .other(String) to force an update for something as major as a new mechanism, basically making sure the driver supports it correctly?

we follow the `.other` pattern elsewhere, e.g. in `ReadConcern`: https://github.com/mongodb/mongo-swift-driver/blob/973acb25f748d7ca27d131c88b109193b839f803/Sources/MongoSwift/ReadConcern.swift#L20

If a new mechanism is added, we'll get new tests added to the auth spec and a ticket created for support + tests, similarly to what happens for driver options like those you've implemented recently. So I think that will satisfy your concern about making sure the driver supports it correctly.

> Perhaps we could move the enum inside the Credential struct giving us the Credential.Mechanism type?
I'm good with nesting the type if we stick with the separate struct approach 👍

Reply:
I wonder if it would be better to just make this an opaque struct. We seem to be adding new authentication mechanisms somewhat often, and the `.other` API isn't the greatest. Because the mechanisms themselves don't have any customization / associated values and don't need to be switched on, we can do this without any API cost. Also, unlike options that are sent to the server, the driver won't start supporting a new auth mechanism without an update of the driver, so the `.other` case only serves to preserve API compatibility but doesn't provide any forward behavioral compatibility.

e.g.:
```swift
public struct Mechanism {
    public static let gssApi = AuthMechanism(name: "GSSAPI")
    public static let scramSha1 = AuthMechanism(name: "SCRAM-SHA-1")

    private init(name: String) {
        self.name = name
    }
}
```
Then due to the nice syntactic stuff we've been using recently for other types, users can still specify the mechanism with the shorthand syntax if they want:
```swift
Credential(mechanism: .scramSha1)
Credential(mechanism: Credential.Mechanism.scramSha1)
```

Knowing that the syntax is nice even when using structs, I'm starting to get the feeling that if we do need an `.other` case, we _probably_ shouldn't be using an enum. While that strategy allow us to support new cases in the future without breaking API, it does so in an unfortunate way for users. I think that until non-exhaustive enums are available outside of the standard library, we should just stick to static structs for enum-like things that may need new cases in the future. As enums are now, one of their primary value-adds is their exhaustiveness, and if we define them for cases that cannot be exhaustively covered, then I don't think they add much. (I think I've reached my enum hangover...).

A survey of the driver for enums that have a `.other` case yields:
- [`CollectionType`](https://mongodb.github.io/mongo-swift-driver/MongoSwift/Enums/CollectionType.html): This is returned as part of the `CollectionSpecification` model in a `listCollections` response. Going back to 3.4, no value besides `collection` and `view` have been possible here.
    - My proposal: we simply drop the `.other` case due to low risk of new value being introduced. Maybe have to confirm with server team
- [`FullDocument`](https://mongodb.github.io/mongo-swift-driver/MongoSwift/Enums/FullDocument.html): The spec specifically requires that users be able to specify unknown values
    - My proposal: convert to struct with a static `updateLookup`, allow `FullDocument` to be initializable from `String` (and string literal)
- [`ReadConcern.Level`](https://mongodb.github.io/mongo-swift-driver/MongoSwift/Structs/ReadConcern/Level.html): New variants are added somewhat regularly (`snapshot` in 4.0 being the most recent)
   - My proposal: convert to struct with static values for the various levels, allow `Level` to be initializable from `String` (and string literal)

Reply:
For my comments not related to this PR, I filed [SWIFT-827](https://jira.mongodb.org/browse/SWIFT-827)